src/djangooidc/views.py

```diff
@@ -0,0 +1,120 @@
+            },
+            status=401,
+        )
+    if isinstance(error, o_e.InternalError):
+        return render(
+            request,
+            "500.html",
+            context={
+                "friendly_message": error.friendly_message,
+                "log_identifier": error.locator,
+            },
+            status=500,
+        )
+    if isinstance(error, Exception):
+        return render(request, "500.html", status=500)
+
+
+def openid(request):
+    """Redirect the user to an authentication provider (OP)."""
+    request.session["next"] = request.GET.get("next", "/")
+
+    try:
+        return CLIENT.create_authn_request(request.session)
+    except Exception as err:
+        return error_page(request, err)
+
+
+def login_callback(request):
+    """Analyze the token returned by the authentication provider (OP)."""
+    try:
+        query = parse_qs(request.GET.urlencode())
+        userinfo = CLIENT.callback(query, request.session)
+        user = authenticate(request=request, **userinfo)
+        if user:
+            login(request, user)
+            logger.info("Successfully logged in user %s" % user)
+            return redirect(request.session.get("next", "/"))
+        else:
+            raise o_e.BannedUser()
+    except Exception as err:
+        return error_page(request, err)
+
+
+def logout(request, next_page=None):
+    """Redirect the user to the authentication provider (OP) logout page."""
+    try:
+        username = request.user.username
+        request_args = {
+            # it is perfectly fine to send the token, even if it is expired
+            "id_token_hint": request.session["id_token_raw"],
+            "state": request.session["state"],
+        }
+        if (
+            "post_logout_redirect_uris" in CLIENT.registration_response.keys()
+            and len(CLIENT.registration_response["post_logout_redirect_uris"]) > 0
+        ):
+            request_args.update(
+                {
+                    "post_logout_redirect_uri": CLIENT.registration_response[
+                        "post_logout_redirect_uris"
+                    ][0]
+                }
+            )
+
+        url = CLIENT.provider_info["end_session_endpoint"]
+        url += "?" + urlencode(request_args)
+        return HttpResponseRedirect(url)
+    except Exception as err:
+        return error_page(request, err)
+    finally:
+        # Always remove Django session stuff - even if not logged out from OP.
+        # Don't wait for the callback as it may never come.
+        auth_logout(request)
+        logger.info("Successfully logged out user %s" % username)
+        next_page = getattr(settings, "LOGOUT_REDIRECT_URL", None)
+        if next_page:
+            request.session["next"] = next_page
+
+
+def logout_callback(request):
+    """Simple redirection view: after logout, redirect to `next`."""
+    next = request.session["next"] if "next" in request.session.keys() else "/"
+    return redirect(next)
```

Comment:
**thought**: I have had problems not getting to the `next` page that I expected. I wonder if we can do any testing of this callback to ensure that it is working as expected.

Reply:
I appreciate you asking for code tests. It will take a bit of time, but I will add them.

I'm unsure if it is possible to do arbitrary `next` redirection with Login.gov, due to the security implications. I'll have to do a bit of reading and experimentation.